flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateSerializer.java

```diff
@@ -0,0 +1,199 @@
+				return !bufferBuilder.isFull();
+			}
+
+			@Override
+			public int writableBytes() {
+				return bufferBuilder.writableBytes();
+			}
+
+			@Override
+			public int writeBytes(InputStream input, int upToBytes) throws IOException {
+				int left = upToBytes;
+				for (int toRead = getToRead(left); toRead > 0; toRead = getToRead(left)) {
+					int read = input.read(buf, 0, toRead);
+					int copied = bufferBuilder.append(java.nio.ByteBuffer.wrap(buf, 0, read));
+					Preconditions.checkState(copied == read);
+					left -= read;
+				}
+				bufferBuilder.commit();
+				return upToBytes - left;
+			}
+
+			private int getToRead(int bytesToRead) {
+				return min(bytesToRead, min(buf.length, writableBytes()));
+			}
+		};
+	}
+
+	static ChannelStateByteBuffer wrap(byte[] bytes) {
+		int written = 0;
+		return new ChannelStateByteBuffer() {
+			@Override
+			public boolean isWritable() {
+				return written < bytes.length;
+			}
+
+			@Override
+			public int writableBytes() {
+				return bytes.length - written;
+			}
+
+			@Override
+			public int writeBytes(InputStream input, int bytesToRead) throws IOException {
+				return input.read(bytes, written, writableBytes());
+			}
+		};
+	}
+}
+
+class ChannelStateSerializerImpl implements ChannelStateSerializer, ChannelStateDeserializer {
+	private static final int LEN_SIZE = Integer.BYTES;
+	private static final int SERIALIZATION_VERSION = 0;
+
+	@Override
+	public void writeHeader(DataOutputStream dataStream) throws IOException {
+		dataStream.writeInt(SERIALIZATION_VERSION);
+	}
+
+	@Override
+	public void writeData(DataOutputStream stream, Buffer... flinkBuffers) throws IOException {
+		stream.writeInt(getSize(flinkBuffers));
+		for (Buffer buffer : flinkBuffers) {
+			ByteBuf nettyByteBuf = buffer.asByteBuf();
+			nettyByteBuf.getBytes(nettyByteBuf.readerIndex(), stream, nettyByteBuf.readableBytes()); // todo: review me
+		}
+	}
+
+	private int getSize(Buffer[] buffers) {
+		int len = 0;
+		for (Buffer buffer : buffers) {
+			len = addExact(len, buffer.readableBytes());
+		}
+		return len;
+	}
+
+	@Override
+	public void readHeader(InputStream stream) throws IOException {
+		int version = readInt(stream);
+		Preconditions.checkArgument(version == SERIALIZATION_VERSION, "unsupported version: " + version);
+	}
+
+	/**
+	 * Reads the length of state.
+	 *
+	 * @return size of state in bytes and offset of the current position resulted from reading.
+	 */
+	@Override
+	public Tuple2<Integer, Integer> readLength(InputStream stream) throws IOException {
+		int len = readInt(stream);
+		Preconditions.checkArgument(len >= 0, "negative state size");
+		return Tuple2.of(len, LEN_SIZE);
+	}
+
+	@Override
+	public int readData(InputStream stream, ChannelStateByteBuffer buffer, int bytes) throws IOException {
+		return buffer.writeBytes(stream, bytes);
+	}
+
+	private static int readInt(InputStream stream) throws IOException {
+		return new DataInputStream(stream).readInt();
```

Comment:
Could we avoid creating the DIS adhoc? You need to read header anyways, so why not always create a DIS and pass it everywhere?

Reply:
It's doesn't do anything except creating this object and this object doesn't escape the method. So I don't think there is any performance issue here (given this is IO on recovery).
Alternatively, I'd just replicate `DataInputStream.readInt` logic.
I'd like to avoid an extra field just to get `readInt` method.
WDYT?

Reply:
I was more thinking to always wrap InputStream in DIS and change the signatures accordingly. Then it would be the same number of fields.
If that is not doable, then please leave as is, don't have a custom `readInt`.